In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv('../../data/raws/WA_Fn-UseC_-Telco-Customer-Churn.csv')

In [2]:
print(df.head())

   customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract Pape

In [3]:
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [5]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [6]:
print("Data duplicate:", df.duplicated().sum())

Data duplicate: 0


Explain each Column

    -Customer ID

    -gender = male / female

    -SeniorCitizen = Whether the customer is a senior citizen or not (1, 0) => ผู้สูงอายุ คือ อาจจะเกษียณแล้ว

    -Partner = Whether the customer has a partner or not (Yes, No)

    -Dependents = Whether the customer has dependents or not (Yes, No)

    -tenure = Number of months the customer has stayed with the company

    -PhoneService = Whether the customer has a phone service or not (Yes, No)

    -MultipleLines = Whether the customer has multiple lines or not (Yes, No, No phone service)

    -InternetService = Customer’s internet service provider (DSL, Fiber optic, No)

    -OnlineSecurity = Whether the customer has online security or not (Yes, No, No internet service)

    -OnlineBackup = Whether the customer has online backup or not (Yes, No, No internet service)

    -DeviceProtection = Whether the customer has device protection or not (Yes, No, No internet service)

    -TechSupport = Whether the customer has tech support or not (Yes, No, No internet service)

    -StreamingTV = Whether the customer has streaming TV or not (Yes, No, No internet service) 

    -StreamingMovies = Whether the customer has streaming movies or not (Yes, No, No internet service)

    -Contract = The contract term of the customer (Month-to-month, One year, Two year) => ลูกค้าใช้ระบบใบแจ้งหนี้แบบไม่ใช้กระดาษหรือไม่?

    -PaymentMethod = The customer’s payment method (Electronic check, Mailed check, Bank transfer (automatic), Credit card)

    -MonthlyCharges = The amount charged to the customer monthly

    - TotalCharges = The total amount charged to the customer

    -Churn = Whether the customer churned or not (Yes or No)

In [7]:
# Check unique values for categorical columns with limited ranges
print("=" * 80)
print("CATEGORICAL COLUMNS VALIDATION")
print("=" * 80)

# Define expected unique values for each categorical column
expected_values = {
    'gender': {'Male', 'Female'},
    'SeniorCitizen': {0, 1},
    'Partner': {'Yes', 'No'},
    'Dependents': {'Yes', 'No'},
    'PhoneService': {'Yes', 'No'},
    'MultipleLines': {'Yes', 'No', 'No phone service'},
    'InternetService': {'DSL', 'Fiber optic', 'No'},
    'OnlineSecurity': {'Yes', 'No', 'No internet service'},
    'OnlineBackup': {'Yes', 'No', 'No internet service'},
    'DeviceProtection': {'Yes', 'No', 'No internet service'},
    'TechSupport': {'Yes', 'No', 'No internet service'},
    'StreamingTV': {'Yes', 'No', 'No internet service'},
    'StreamingMovies': {'Yes', 'No', 'No internet service'},
    'Contract': {'Month-to-month', 'One year', 'Two year'},
    'PaperlessBilling': {'Yes', 'No'},
    'PaymentMethod': {'Electronic check', 'Mailed check', 'Bank transfer (automatic)', 'Credit card (automatic)'},
    'Churn': {'Yes', 'No'}
}

# Check each categorical column
for column, expected in expected_values.items():
    if column in df.columns:
        actual = set(df[column].unique())
        print(f"\n✓ {column}:")
        print(f"  Expected values: {sorted([str(x) for x in expected])}")
        print(f"  Actual values:   {sorted([str(x) for x in actual])}")
        
        # Check for unexpected values
        unexpected = actual - expected
        if unexpected:
            print(f"  ⚠️  UNEXPECTED VALUES: {unexpected}")
        
        # Show value counts
        print(f"  Value counts:")
        print(f"  {df[column].value_counts().to_dict()}")
    else:
        print(f"\n✗ {column}: NOT FOUND in dataset")

print("\n" + "=" * 80)
print("NUMERICAL COLUMNS SUMMARY")
print("=" * 80)
print(df[['tenure', 'MonthlyCharges', 'TotalCharges']].describe())

CATEGORICAL COLUMNS VALIDATION

✓ gender:
  Expected values: ['Female', 'Male']
  Actual values:   ['Female', 'Male']
  Value counts:
  {'Male': 3555, 'Female': 3488}

✓ SeniorCitizen:
  Expected values: ['0', '1']
  Actual values:   ['0', '1']
  Value counts:
  {0: 5901, 1: 1142}

✓ Partner:
  Expected values: ['No', 'Yes']
  Actual values:   ['No', 'Yes']
  Value counts:
  {'No': 3641, 'Yes': 3402}

✓ Dependents:
  Expected values: ['No', 'Yes']
  Actual values:   ['No', 'Yes']
  Value counts:
  {'No': 4933, 'Yes': 2110}

✓ PhoneService:
  Expected values: ['No', 'Yes']
  Actual values:   ['No', 'Yes']
  Value counts:
  {'Yes': 6361, 'No': 682}

✓ MultipleLines:
  Expected values: ['No', 'No phone service', 'Yes']
  Actual values:   ['No', 'No phone service', 'Yes']
  Value counts:
  {'No': 3390, 'Yes': 2971, 'No phone service': 682}

✓ InternetService:
  Expected values: ['DSL', 'Fiber optic', 'No']
  Actual values:   ['DSL', 'Fiber optic', 'No']
  Value counts:
  {'Fiber optic': 30

In [8]:
#Count churn amount
churn_counts = df['Churn'].value_counts()
print("\nChurn Value Counts:")
print(churn_counts)


Churn Value Counts:
Churn
No     5174
Yes    1869
Name: count, dtype: int64


In [9]:
# Check churn distributione
churn_imbalance_ratio = churn_counts['No'] / churn_counts['Yes']
print(f"\nChurn Imbalance Ratio : {churn_imbalance_ratio:.2f}")

# Check churn distribution to percentage
churn_rate = churn_counts['Yes'] / churn_counts.sum() * 100
print(f"Churn Rate : {churn_rate:.2f}%")


Churn Imbalance Ratio : 2.77
Churn Rate : 26.54%


#### All Data Clean and not have any duplicate data we can Feature engineering and analyst

### ภาพรวม ================================

จํานวนลูกค้า : 7043

จํานวน churn : No 5174 / Yes    1869

Churn rate = 26.54%

Churn Ratio = 2.77

Dataset contains 7,043 customers with a churn rate of approximately 26%, meaning around 1,800 customers discontinued the service. This represents a significant business impact and justifies the need for predictive modeling.

In [10]:
# Save cleaned data for next step
df.to_csv('../../data/clean/cleaned_telco_churn.csv', index=False)